cpp/src/arrow/c/bridge.cc

```diff
@@ -2250,101 +2336,126 @@ class ArrayStreamReader {
     return {code, last_error ? std::string(last_error) : ""};
   }
 
+  DeviceAllocationType get_device_type() const {
+    if constexpr (std::is_same_v<ArrayType, struct ArrowDeviceArray>) {
+      return static_cast<DeviceAllocationType>(stream_.device_type);
+    } else {
+      return DeviceAllocationType::kCPU;
+    }
+  }
+
  private:
-  mutable struct ArrowArrayStream stream_;
+  mutable StreamType stream_;
+  const DeviceMemoryMapper& mapper_;
```

Comment:
Is it ok to keep a mere reference here?

Reply:
it *should* be okay since it's an `std::function`, I wanted to avoid performing a copy if it wasn't necessary.

Reply:
I don't get why it being a `std::function` makes it ok. A `std::function` can own arbitrary state (for example a closure's captured variables).

Reply:
Would be nice to be able to enforce that only move-only functions [1] be accepted here.

[1] https://github.com/ofats/any_invocable

Reply:
Let's not overdo this either.

Reply:
> Let's not overdo this either.

So just removing the `&` is what you have in mind?

Reply:
Yes, exactly.

Reply:
After doing a bit of research, not only is @pitrou correct that the current version isn't necessarily safe, but in most cases it's preferable and more performant to pass an `std::function` by value and use `std::move` instead. So I'll update this.